### ISSUE TITLE
Expected behavior when unexpected event is received?

### ISSUE THREAD

**Issue by cgillum** [Azure/azure-functions-durable-extension#29]
## Question
Unexpected external events are handled by orchestrator functions don't result in any actual work done, and append to the execution history. Is this behavior acceptable?

## Example
Consider the `E3_Counter` example function:

```csharp
[FunctionName("E3_Counter")]
public static async Task<int> Run(
    [OrchestrationTrigger] DurableOrchestrationContext counterContext)
{
    int counterState = counterContext.GetInput<int>();
    string operation = await counterContext.WaitForExternalEvent<string>("operation");

    operation = operation?.ToLowerInvariant();
    if (operation == "incr")
    {
        counterState++;
    }
    else if (operation == "decr")
    {
        counterState--;
    }

    if (operation != "end")
    {
        counterContext.ContinueAsNew(counterState);
    }

    return counterState;
}
```

The orchestrator function starts and waits for an event named `operation`. If it instead receives an event named `bogus`, it replays and stops at the `WaitForExternalEvent` call.  This can get repeated seveal times and the history gets updated similar to the following:

Row Key  |  Event Type
-- | --
0000000000000000 | OrchestratorStarted
0000000000000001 | ExecutionStarted
0000000000000002 | OrchestratorCompleted
0000000000000003 | OrchestratorStarted
0000000000000004 | EventRaised
0000000000000005 | OrchestratorCompleted
0000000000000006 | OrchestratorStarted
0000000000000007 | EventRaised
0000000000000008 | OrchestratorCompleted
0000000000000009 | OrchestratorStarted
000000000000000A | EventRaised
000000000000000B | OrchestratorCompleted

As can be seen, there is a new episode of event history for each discarded event taking up storage space and potentially impacting replay execution time.

What can/should be done about this? Are there any other adverse affects that have not been considered?

**Comment by cgillum** [comment]
So far, this has not been an issue that folks have raised, so probably safe to close this issue.

**Comment by mrjesseethompson** [comment]
Hi there. We *were* trying out durable functions (the Eternal Orchestrations pattern) and were disconcerted when we discovered that calling `.RaiseEventAsync()` throws away events if the orchestrator is busy handling other events, or if it is still starting up.  I got to this page by following the link [here](https://docs.microsoft.com/en-us/azure/azure-functions/durable-functions-external-events) BTW, which states:

```
Warning

If there is no orchestration instance with the specified instance ID or if the instance is not waiting on the specified event name, the event message is discarded. For more information about this behavior, see the GitHub issue.
```

Monitoring the status using `.GetStatusAsync` didn't help us much because `DurableOrchestrationStatus.RuntimeStatus == OrchestrationRuntimeStatus.Running` before `.WaitForExternalEvent<Event>("EventName")` is invoked (at least during local development).

Even if we could reliably check that the orchestrator is already running, we receive a set of events (from a Service Bus trigger) in fairly rapid succession, so when we forward those to the Orchestrator it is highly likely that its busy processing the last one that it received (not currently `await`ing `WaitForExternalEvent()`).

Here's the full explanation of our use-case.

1) We receive a stream of messages from Service Bus. Each message contains information about the progress of an executing process in an external system. We're not leveraging Service Bus sessions, so these events can arrive in any order, but events do have a `correlationId`, and at some point we expect to see a completion event.
2) When we receive a message, we use the` correlationId` as the identifier of a new Orchestration Function instance that will be used to track the overall state of the external system. We create the Orchestrator if it does not yet exist.
3) Upon creation, the Orchestration function:

i. Waits for an event
ii. Updates its state to track the overall progress of the external system using the new information gained from the event.
iii. Depending on the state of the external process, it may then invoke other functions to start downstream processing of the output from the external system, without having to wait for it to run to completion.
iv. Restarts itself using `.ContinueAsNew()`
v. Waits for the next event.

4) Finally, when the external system has completed, we finish up and let the Orchestrator shut-down.

In summary, we can't help but feel confused about how to use Durable Function's eventing in any useful application. It appears that the only scenario where it could work reliably is if we expected to receive exactly one event at a time fairly far in the future, which seems like a niche scenario.

Anyway, I'd appreciate your feedback. We've moved on to a different solution obviously, but continue to keep an eye on what's happening with server-less tech. Thanks!

**Comment by cgillum** [comment]
Thanks for taking the time to post this in spite of moving to a different solution. The mistake on our part was advertising the "actor" pattern early in the preview, where these types of messaging patterns are most common.

One specific issue it appears you ran into is this: [Message loss due to race condition with ContinueAsNew](https://github.com/Azure/azure-functions-durable-extension/issues/67). To summarize, it turns out that `ContinueAsNew` temporarily puts the orchestrator into a limbo state and any events received while in the state will be dropped. If *not* using `ContinueAsNew`, the orchestrator can handle multiple external events concurrently and won't drop any of them. We've updated our docs and samples to remove references to actor patterns and de-emphasize `ContinueAsNew`, but that happened only recently.

So again, just to clarify, multiple events can be received at the same time just fine. The problem is that 1) the instance must be running and 2) the instance must not be in the `ContinueAsNew` state. We have work items to clarify this further in the documentation before we go into GA. Again, thanks for your detailed feedback - it's very helpful.

**Comment by kawazoe** [comment]
@cgillum Is this still an issue with the fixes from the 1.8.0 release?

I was hoping to use events with continue as new to implement a debounce pattern similar to what was discussed by @mrjesseethompson a year ago. My plan was to send some correlated messages body as events until I get the last message of the session. At this point, the WaitForExternalEvent call would time out and the orchestrator would begin its work with the most recent event as input. I used the same ContinueAsNew pattern between each event to extend the debounce timeout as I would get more messages.
